server/coordinator/procedure/manager_impl.go

```diff
@@ -123,34 +122,62 @@ func (m *ManagerImpl) retryAll(ctx context.Context) error {
 	return nil
 }
 
-func (m *ManagerImpl) startProcedureWorker(ctx context.Context, procedures <-chan Procedure) {
-	for procedure := range procedures {
-		err := procedure.Start(ctx)
-		if err != nil {
-			log.Error("procedure start failed", zap.Error(err))
+func (m *ManagerImpl) startProcedurePromote(ctx context.Context) {
+	ticker := time.NewTicker(defaultPromoteDelay)
+	procedureWorkerChan := make(chan bool)
+	for {
+		select {
+		case <-ticker.C:
+			m.startProcedurePromoteInternal(ctx, procedureWorkerChan)
+		case <-procedureWorkerChan:
+			m.startProcedurePromoteInternal(ctx, procedureWorkerChan)
+		case <-ctx.Done():
+			return
 		}
-
-		m.removeProcedure(procedure.ID())
 	}
 }
 
-func (m *ManagerImpl) removeProcedure(id uint64) Procedure {
+func (m *ManagerImpl) startProcedurePromoteInternal(ctx context.Context, procedureWorkerChan chan bool) {
+	newProcedures, err := m.promoteProcedure(ctx)
+	if err != nil {
+		log.Error("promote procedure failed", zap.Error(err))
+		return
+	}
+
 	m.lock.Lock()
 	defer m.lock.Unlock()
-
-	index := -1
-	for i, p := range m.procedures {
-		if p.ID() == id {
-			index = i
-			break
+	for _, newProcedure := range newProcedures {
+		log.Info("promote procedure", zap.Uint64("procedureID", newProcedure.ID()))
```

Comment:
How about logging this with `startProcedureWorker` in another forloop outside the lock.